File: zig/src/journal.zig
const util = @import("util.zig");
const types = @import("types.zig");

pub fn j2ffs(self: anytype) void {
    const main = self.fp orelse return;
    const fpj = std.Io.Dir.cwd().openFile(self.io, self.journal[0..self.journal_len], .{}) catch return;
    defer {
        fpj.close(self.io);
        util.deleteFileIfExists(self.io, self.journal[0..self.journal_len]);
    }

    var state: [1]u8 = undefined;
    util.readExactAt(self.io, fpj, 0, state[0..]) catch return;
    if (state[0] != 0xff) {
        return;
    }

    var index_block: [4 + types.BlockSize]u8 = [_]u8{0} ** (4 + types.BlockSize);
    var pos: u64 = 1;
    while (true) {
        util.readExactAt(self.io, fpj, pos, index_block[0..]) catch break;
        const index = util.b4ToU32(index_block[0..4]);
        util.writeAllAt(self.io, main, @as(u64, index) * @as(u64, types.BlockSize), index_block[4..]) catch break;
        pos += index_block.len;
    }
    util.fflush(self.io, main);
}

const std = @import("std");

File: cpp/src/journal.cpp
#include "types.hpp"

#include "util.hpp"

namespace filefs::detail {

void FileSystemState::j2ffs() {
  CFile journal_fp;
  journal_fp.reset(std::fopen(fnj.c_str(), "rb"));
  if (!journal_fp) {
    return;
  }

  const auto cleanup = [this, &journal_fp]() noexcept {
    journal_fp.reset();
    remove_if_exists(fnj);
  };

  std::uint8_t state = 0U;
  if (!read_exact(journal_fp.get(), &state, 1U) || state != 0xffU) {
    cleanup();
    return;
  }

  std::array<std::uint8_t, BlockSize + 4> index_block{};
  while (read_exact(journal_fp.get(), index_block.data(), index_block.size())) {
    const std::uint32_t index = b4_to_u32(std::span<const std::uint8_t, 4>(index_block.data(), 4));
    const std::uint64_t pos = static_cast<std::uint64_t>(index) * BlockSize;
    if (!set_pos(fp.get(), pos) || !write_exact(fp.get(), index_block.data() + 4, BlockSize)) {
      break;
    }
  }
  flush_file(fp.get());
  cleanup();
}

}  // namespace filefs::detail
